automates/model_analysis/identification_algorithms.py

```diff
@@ -0,0 +1,251 @@
+    :param stop_on_noid: should causal_effect halt when non-identifiability is determined?
+    :return: P(y |do(x), z) in LaTeX code
+    """
+    if "description" not in g.edge_attributes():
+        g.es["description"] = numpy.repeat("O", len(g.es))
+    g_obs = gm.observed_graph(g)
+    if not g_obs.is_dag():
+        raise ValueError("Graph 'G' is not a DAG.")
+    topo_ind = g_obs.topological_sorting()
+    topo = gm.to_names(topo_ind, g_obs)
+    if len(set(y) - set(topo)) > 0:
+        raise ValueError("Set 'y' contains variables not present in the graph.")
+    if len(set(x) - set(topo)) > 0:
+        raise ValueError("Set 'x' contains variables not present in the graph.")
+    if z is not None:
+        if len(set(z) - set(topo)) > 0:
+            raise ValueError("Set 'z' contains variables not present in the graph.")
+    if len(set(x) & set(y)) > 0:
+        raise ValueError("Sets 'x' and 'y' are not disjoint.")
+    if z is not None:
+        if len(set(y) & set(z)) > 0:
+            raise ValueError("Sets 'y' and 'z' are not disjoint.")
+        if len(set(x) & set(z)) > 0:
+            raise ValueError("Sets 'x' and 'z' are not disjoint.")
+    if z is None:
+        res = compute_ID(y, x, gm.Probability(), g, g_obs, topo, topo, gm.TreeNode())
+        algo = "id"
+        res_prob = res.p
+    else:
+        res = compute_IDC(y, x, z, gm.Probability(), g, g_obs, topo, topo, gm.TreeNode())
+        algo = "idc"
+        res_num = deepcopy(res.p)
+        res_den = deepcopy(res_num)
+        res_den.sumset = list(set(y) | set(res_den.sumset))
+        res_prob = gm.Probability(fraction=True)
+        res_prob.num = deepcopy(res_num)
+        res_prob.den = deepcopy(res_den)
+
+    res_tree = res.tree
+    if res.tree.call.id_check:
+        output = gm.Results(query={"y": y, "x": x, "z": z}, algorithm=algo, p=gm.get_expression(res_prob),
+                            tree=res_tree)
+        if steps:
+            return output
+        return output.p
+    else:
+        if stop_on_noid:
+            raise gm.IDANotIdentifiable("Not Identifiable")
+        output = gm.Results(query={"y": y, "x": x, "z": z}, algorithm=algo, p="", tree=res_tree)
+        if steps:
+            return output
+        return output.p
+
+
+def compute_ID(y, x, p, g, g_obs, v, topo, tree):
+    """
+    Performs Shpitser and Pearl's ID algorithm
+    :param y: value assignment outcome
+    :param x: value assignment intervention
+    :param p: probability object
+    :param g: full graph
+    :param g_obs: observed graph
+    :param v: vertices
+    :param topo: topological ordering
+    :param tree: object to store line-by-line summary
+    :return: P(y | do(x)) if y is identifiable or failure.
+                Note that the output needs to be read by get_expression
+    """
+    to = None
+    frm = None
+    description = None
+    tree = gm.TreeNode()
+    if (len(p.var) == 0) and (not (p.product or p.fraction)):
```

Comment:
@aswinchester ideally we want as little code duplication under conditionals as possible. Consider the assignments under this conditional. They have quite a lot in common and only a single difference. The following simplification will make the code a lot easier to read:

```python
if (len(p.var) == 0) and (not (p.product or p.fraction)):
  p = gm.Probability(var=v)

tree.call = gm.Call(y=y, x=x, p=p, g=g, line=0, v=v, id_check=False)
```
Notice that when we use this method of assigning the minimal number of constructs that need to be unique under a conditional we can actually eliminate the `else` branch entirely. Please make this change in your code.

Reply:
This has been completed